pkg/xds/experimental/client/client_test.go

```diff
@@ -0,0 +1,1116 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"slices"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	clusterpb "github.com/cilium/proxy/go/envoy/config/cluster/v3"
+	corepb "github.com/cilium/proxy/go/envoy/config/core/v3"
+	endpointpb "github.com/cilium/proxy/go/envoy/config/endpoint/v3"
+	listenerpb "github.com/cilium/proxy/go/envoy/config/listener/v3"
+	routepb "github.com/cilium/proxy/go/envoy/config/route/v3"
+	discoverypb "github.com/cilium/proxy/go/envoy/service/discovery/v3"
+	durationpb "google.golang.org/protobuf/types/known/durationpb"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/cilium/cilium/pkg/envoy"
+)
+
```

Comment:
I would suggesting using https://github.com/uber-go/goleak in the tests to verify that all background goroutines are cleaned up when stopping. This does require that they are properly waited for.